docs/RFCS/20201020_sql_query_denylist.md

```diff
@@ -0,0 +1,183 @@
+
+### Configuring denylist location
+A denylist location flag can be configured in [pkg/cli/flags.go](https://github.com/cockroachdb/cockroach/blob/master/pkg/cli/flags.go). eg. `--denylist=’denylist.yaml’`. If this flag has not been set, then there will be no location to watch for a denylist and therefore the functionality is off by default. 
+
+### File-managed denylist implementation
+We expect users to create their own YAML file with a `sql` field that lists an array of strings representing regex to match any SQL queries to be denied. This allows granular control of specific command types as well as database objects — a user should be able to denylist any command of a certain type, any command pertaining to a particular database object, or any additional parameters that may be affecting the cluster.
+
+In the future, we could extend the functionality with additional fields such as, for example, `IP` to support denying queries made from specific IP addresses. 
+
+Example denylist: **denylist.yaml**
+```
+# Queries to be denied; in the future we may support denying other fields.
+sql:
+   # Denies the experimental temp tables feature to be turned on. 
+   - “SET experimental_enable_temp_tables=on”
+   # Denies creating any temporary table.
+   - “CREATE TEMP TABLE .*”
+   # Denies creating a changefeed for any table.
+   - “CREATE CHANGEFEED FOR .*”
+   # Denies creating any backups for only database db1.   
+   - “BACKUP DATABASE db1 .*”
+```
+
+If the denylist location flag has been set, an [fsnotify](https://pkg.go.dev/github.com/fsnotify/fsnotify) watcher will be registered when the server is initialized. fsnotify’s built-in functionality will be used to check if a file in the desired location has been added, modified, or removed. If it has been added or modified, the entire YAML file will be read into a YAML Go Struct:
+
+```
+type denylistImpl struct {
+   SQL []string `yaml:"sql,flow"`
+}
+``` 
+
+Each query to be denied from the YAML file will then be updated in real-time in a slice of queries in a newly defined denylist package; every time the file in this location is modified, the entire file will be read again, parsed, and added to the slice of queries. Additionally, every time the location is modified, there will be logging output to track how often the query denylist is modified. 
+
+In the denylist package, this private slice of denylist queries is stored as strings representing regular expressions. The slice of denylist queries needs to be stored in an [`atomic.Value`](https://tip.golang.org/pkg/sync/atomic/#Value) to prevent races against the watcher overwriting the config: each SQL query that comes in must be compared against an atomic slice and further writes to this copy of the slice (from modifications to the denylist YAML file) must be prevented. 
```

Comment:
does `atomic.Value` add a single point of contention for all incoming SQL queries? i'm not sure how it works internally, but we'd want something that allows multiple concurrent readers, or one writer. or maybe we could think of another way to reduce contention by allowing some read-staleness of this slice?

Reply:
ah i see RwMutex being mentioned later. why not use that in the intial design?

well i guess thinking a bit more, the only think we really care about is writes not being lost. it doesn't matter much if reads are kind of out of date. so i'm sure we can come up with a design with less contention.

Reply:
(i think atomic.Value is cheaper than RWMutex as it does not block -- i don't think there's contention just a slightly more expensive CPU instructions)

Reply:
i see, thanks! i had a very incorrect understanding of things.